### ISSUE TITLE
Assembly cache bug with mixed spaces

### ISSUE THREAD

**Issue by dorugeber** [firedrakeproject/firedrake#541]
Couldn't find this issue in the other assembly cache bug threads.  Normally I'm doing a linear solve for the mixed function inside a timestepping loop, but in this minimal example I increment the dat.data directly.

```
from firedrake import *

#parameters["assembly_cache"]["enabled"] = False

mesh = UnitSquareMesh(1, 1)

U = FunctionSpace(mesh, "CG", 2)
V = FunctionSpace(mesh, "CG", 1)
W = U*V

fg = Function(W)
f, g = split(fg)
f_form = f*f*dx
g_form = g*g*dx

print assemble(f_form), assemble(g_form)

for ii in range(5):
    fg.sub(0).dat.data[:] += 1.0
    fg.sub(1).dat.data[:] += 2.0
    print assemble(f_form), assemble(g_form)
```

With the assembly cache:
```
atm112@ubuntu:~$ python assembly_bug.py 
0.0 0.0
0.0 0.0
0.0 0.0
0.0 0.0
0.0 0.0
0.0 0.0
```

Without the assembly cache:
```
atm112@ubuntu:~$ python assembly_bug.py 
0.0 0.0
1.0 4.0
4.0 16.0
9.0 36.0
16.0 64.0
25.0 100.0
```

**Comment by wence-** [comment]
So the reason for this one is breakage in dat versioning.

The dependencies in the assembly cache record the version of the form dependencies, which are the coordinate field and the /mixed/ coefficient fg.  But the two steps don't ever change the version of the mixed coefficient, instead they change the version of the dats inside it.

The simple option is to fix the dep tracking.  The other one is to fix the version property that a mixeddat exposes.  Here's a patch that does the former:


```patch
diff --git a/firedrake/assembly_cache.py b/firedrake/assembly_cache.py
index 3eedced..cbb7c68 100644
--- a/firedrake/assembly_cache.py
+++ b/firedrake/assembly_cache.py
@@ -51,6 +51,8 @@ try:
 except (ImportError, AttributeError):
     memory = None
 
+# Track version of all sub-dats as well as us
+version = lambda dep: tuple([dep.dat._version] + [d._version for d in dep.dat])
 
 class _DependencySnapshot(object):
     """Record the dependencies of a form at a particular point in order to
@@ -60,7 +62,7 @@ class _DependencySnapshot(object):
 
         # For each dependency, we store a weak reference and the
         # current version number.
-        ref = lambda dep: (weakref.ref(dep), dep.dat._version)
+        ref = lambda dep: (weakref.ref(dep), version(dep))
 
         deps = []
 
@@ -76,10 +78,11 @@ class _DependencySnapshot(object):
         """Check whether form is valid with respect to this dependency snapshot."""
 
         original_coords = self.dependencies[0][0]()
+
         if original_coords:
             coords = form.integrals()[0].domain().data().coordinates
             if coords is not original_coords or \
-               coords.dat._version != self.dependencies[0][1]:
+               version(coords) != self.dependencies[0][1]:
                 return False
         else:
             return False
@@ -91,7 +94,7 @@ class _DependencySnapshot(object):
         for original_d, dep in zip(self.dependencies[1:], deps):
             original_dep = original_d[0]()
             if original_dep:
-                if dep is not original_dep or dep.dat._version != original_d[1]:
+                if dep is not original_dep or version(dep) != original_d[1]:
                     return False
             else:
                 return False
```

**Comment by wence-** [comment]
FWIW, I am in favour of turning assembly caching *off* by default, because there have been a number of bugs hanging around unfixed for over 6 months now and they normally exhibit in highly non-trivial to catch ways.

**Comment by dorugeber** [comment]
But then it might as well not be there at all, right?  If the user goes to the trouble of turning on assembly caching, they can go to the trouble of setting up a LinearVariationalSolver or LinearSolver object.

**Comment by wence-** [comment]
there are places where it exhibits in addition to those places (e.g. norm computation of subfields or similar, as you're experiencing), however, I think generally the firedrake side of the caching is basically fine, it's just the pyop2 dat versioning and copying is not perfect.

**Comment by wence-** [comment]
Fixed and tested as part of #555
